### ISSUE TITLE
Optional parameters in TypeScript may lead to compile error

### ISSUE THREAD

**Issue by wapolinar** [NSwag/NSwag#28]
Thanks again for your quick response.

After the introduction of optional parameters, sometimes invalid TypeScript is generated, for example:
````
{
  "swagger" : "2.0",
  "info" : {
    "version" : "1.0.2",
    "title" : "Test API"
  },
  "host" : "localhost:8080",
  "basePath" : "/",
  "tags" : [ {
    "name" : "api"
  } ],
  "schemes" : [ "http" ],
  "paths" : {
     "/removeElement" : {
      "delete" : {
        "tags" : [ "api" ],
        "summary" : "Removes elements",
        "description" : "Removes elements",
        "operationId" : "removeElement",
        "consumes" : [ "application/json" ],
        "produces" : [ "application/json" ],
        "parameters" : [ {
          "name" : "X-User",
          "in" : "header",
          "description" : "User identifier",
          "required" : true,
          "type" : "string"
        }, {
          "name" : "elementId",
          "in" : "query",
          "description" : "The ids of existing elements that should be removed",
          "required" : false,
          "type" : "integer",
        }, {
          "name" : "nameId",
          "in" : "query",
          "description" : "The name id that should be removed",
          "required" : true,
          "type" : "integer",
        } ],
        "responses" : {
          "default" : {
            "description" : "successful operation"
          }
        }
      }
    }
  },
    "definitions" : { }
}
````
Generates:
````javascript
(...)
/**
     * Removes elements
     * @x_User User identifier
     * @elementId The ids of existing elements that should be removed
     * @nameId The name id that should be removed
     * @return successful operation
     */
    removeElement(x_User: string, elementId?: number, nameId: number, onSuccess?: (result: any) => void, onFail?: (exception: string, reason: string) => void) {
(...)
````
This is invalid in TypeScript since elementId cannot be optional, when followed by a required parameter. The ***?*** behind ***elementId*** should be removed...

If you want to keep the ***"required" : true*** vs. ***"required" : "false"*** semantics, the code generation could be changed in such a way that the lines
````javascript
        if (elementId !== undefined && elementId !== null)
````
for each element are only created, when the parameter was set to optional (***"required" : false***). Of course this is optional and does not directly relate to the issue I am describing here.

**Comment by rsuter** [comment]
I think we should never render the ? (everything is nullable) and always check (if statement). If null is passed for required parameter, then the server should return a 500...

**Comment by rsuter** [comment]
The interface now looks like this: 

	export interface IClient {
		/**
		 * Removes elements
		 * @x_User User identifier
		 * @elementId (optional) The ids of existing elements that should be removed
		 * @nameId The name id that should be removed
		 * @return successful operation
		 */
		removeElement(x_User: string, elementId: number, nameId: number ): Q.Promise<any>;
	}

And you can call `removeElement("foo", null, "bar")` if you want to not set the optional parameter.

**Comment by rsuter** [comment]
Fixed in v1.22

**Comment by wapolinar** [comment]
Yes, this is exactly how I do it at the moment, thanks! :-)
